File: examples/go/angzarr/validation.go
package angzarr

// RequireExists checks that a field is non-empty (entity exists).
func RequireExists(field, errMsg string) *CommandError {
	if field == "" {
		return NewFailedPrecondition(errMsg)
	}
	return nil
}

// RequireNotExists checks that a field is empty (entity does not yet exist).
func RequireNotExists(field, errMsg string) *CommandError {
	if field != "" {
		return NewFailedPrecondition(errMsg)
	}
	return nil
}

// RequirePositive checks that a value is greater than zero.
func RequirePositive(value int32, errMsg string) *CommandError {
	if value <= 0 {
		return NewFailedPrecondition(errMsg)
	}
	return nil
}

// RequireNonNegative checks that a value is zero or greater.
func RequireNonNegative(value int32, errMsg string) *CommandError {
	if value < 0 {
		return NewFailedPrecondition(errMsg)
	}
	return nil
}

// RequireNotEmpty checks that a slice has at least one element.
func RequireNotEmpty[T any](items []T, errMsg string) *CommandError {
	if len(items) == 0 {
		return NewFailedPrecondition(errMsg)
	}
	return nil
}

// RequireStatus checks that the current status matches the expected value.
func RequireStatus(actual, expected, errMsg string) *CommandError {
	if actual != expected {
		return NewFailedPrecondition(errMsg)
	}
	return nil
}

// RequireStatusNot checks that the current status is NOT the forbidden value.
func RequireStatusNot(actual, forbidden, errMsg string) *CommandError {
	if actual == forbidden {
		return NewFailedPrecondition(errMsg)
	}
	return nil
}

File: client/python/angzarr_client/validation.py
"""Validation helpers for command handler precondition checks.

Eliminates repeated validation boilerplate across aggregate handlers.
"""

from collections.abc import Sequence
from typing import Any

from .errors import CommandRejectedError


def require_exists(field: str, error_msg: str) -> None:
    """Require that a field is non-empty (entity exists)."""
    if not field:
        raise CommandRejectedError(error_msg)


def require_not_exists(field: str, error_msg: str) -> None:
    """Require that a field is empty (entity does not yet exist)."""
    if field:
        raise CommandRejectedError(error_msg)


def require_positive(value: int, error_msg: str) -> None:
    """Require that a value is greater than zero."""
    if value <= 0:
        raise CommandRejectedError(error_msg)


def require_non_negative(value: int, error_msg: str) -> None:
    """Require that a value is zero or greater."""
    if value < 0:
        raise CommandRejectedError(error_msg)


def require_not_empty(items: Sequence[Any], error_msg: str) -> None:
    """Require that a sequence has at least one element."""
    if not items:
        raise CommandRejectedError(error_msg)


def require_status(actual: str, expected: str, error_msg: str) -> None:
    """Require that the current status matches the expected value."""
    if actual != expected:
        raise CommandRejectedError(error_msg)


def require_status_not(actual: str, forbidden: str, error_msg: str) -> None:
    """Require that the current status is NOT the forbidden value."""
    if actual == forbidden:
        raise CommandRejectedError(error_msg)
